File: blog-sever/server-kt/src/main/kotlin/com/site/blog/controller/LinkController.kt
package com.site.blog.controller

import com.baomidou.mybatisplus.extension.kotlin.KtQueryWrapper
import com.site.blog.aop.LogOperationEnum
import com.site.blog.aop.SysLogAnnotation
import com.site.blog.constants.HttpStatusEnum
import com.site.blog.constants.LinkConstants
import com.site.blog.model.dto.AjaxPutPage
import com.site.blog.model.dto.AjaxResultPage
import com.site.blog.model.entity.Link
import com.site.blog.service.LinkService
import com.site.blog.util.DateUtils.localCurrentDate
import com.site.blog.util.BaseResult.getResultByHttp
import com.site.blog.util.ResultDto
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.web.bind.annotation.*
import java.sql.Timestamp
import java.time.LocalDateTime

/**
 * @author yzqde
 * @qq 1320291471
 * @Description: 友链Controller
 * @create 2020/12/27
 */
@RestController
@RequestMapping("/v2/admin")
@Tag(name = "链接json")
class LinkController(private val linkService: LinkService) {

    @GetMapping("/linkType/list")
    fun linkTypeList(): ResultDto<List<Link>> {
        val links: MutableList<Link> = ArrayList()
        links.add(
            Link(
                linkType = LinkConstants.LINK_TYPE_FRIENDSHIP.linkTypeId,
                linkName = LinkConstants.LINK_TYPE_FRIENDSHIP.linkTypeName
            )
        )
        links.add(
            Link(
                linkType = LinkConstants.LINK_TYPE_RECOMMEND.linkTypeId,
                linkName = LinkConstants.LINK_TYPE_RECOMMEND.linkTypeName
            )
        )
        links.add(
            Link(
                linkType = LinkConstants.LINK_TYPE_PRIVATE.linkTypeId,
                linkName = LinkConstants.LINK_TYPE_PRIVATE.linkTypeName
            )
        )


        return getResultByHttp(HttpStatusEnum.OK, links)
    }

    /**
     * /v2/admin/link/paging?page=1&limit=30
     *
     * @param ajaxPutPage
     * @return
     */
    @GetMapping("/link/paging")
    fun getLinkList(ajaxPutPage: AjaxPutPage<Link?>?): ResultDto<*> {
        val queryWrapper = KtQueryWrapper(Link())
        queryWrapper
            .orderByAsc(Link::linkRank)
        return if (ajaxPutPage != null) {
            val page = ajaxPutPage.putPageToPage()
            linkService.page(
                page,
                queryWrapper
            )
            val result = AjaxResultPage<Link?>()
            result.list = page.records
            result.count = page.total
            getResultByHttp(
                HttpStatusEnum.OK,
                true,
                result
            )
        } else {
            getResultByHttp(HttpStatusEnum.BAD_REQUEST, false, "fail")
        }
    }

    @PostMapping("/link/hide")
    @SysLogAnnotation(title = "删除链接", opType = LogOperationEnum.CHANGE_STATUS)
    fun updateLinkStatus(link: Link): ResultDto<*> {
        val flag = linkService.updateById(link)
        return if (flag) {
            getResultByHttp(HttpStatusEnum.OK, link)
        } else getResultByHttp(HttpStatusEnum.INTERNAL_SERVER_ERROR)
    }

    @DeleteMapping("/link/clear/{id}")
    @SysLogAnnotation(title = "清除链接", opType = LogOperationEnum.CLEAN)
    fun clearLink(@PathVariable("id") linkId: String): ResultDto<*> {
        val flag = linkService.removeById(linkId)
        return if (flag) {
            getResultByHttp(HttpStatusEnum.OK, true, linkId)
        } else getResultByHttp(HttpStatusEnum.INTERNAL_SERVER_ERROR)
    }

    @PostMapping("/link/edit")
    @SysLogAnnotation(title = "编辑链接", opType = LogOperationEnum.EDIT)
    fun updateAndSaveLink(link: Link): ResultDto<*> {
        try {
            link.createTime = localCurrentDate
            val flag: Boolean
            if (link.linkId?.isNotEmpty() == true) {
                println(link.linkId)
                println("编辑链接->更改")
                link.updateTime = Timestamp.valueOf(LocalDateTime.now())
                flag = linkService.updateById(link)
            } else {
                link.linkRank = 1
                link.show = true
                flag = linkService.save(link)
            }
            if (flag) {
                return getResultByHttp(HttpStatusEnum.OK, link.linkId)
            }
        } catch (e: Exception) {
            e.printStackTrace()
            return getResultByHttp(HttpStatusEnum.INTERNAL_SERVER_ERROR, e.message)
        }
        return getResultByHttp(HttpStatusEnum.INTERNAL_SERVER_ERROR)
    }
}

File: blog-go/controller/link_controller.go
package controller

import (
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"myblog-go/model"
	"myblog-go/util"
)

func LinkTypeList(c *gin.Context) {
	links := []gin.H{
		{"linkType": 0, "linkName": "友情链接"},
		{"linkType": 1, "linkName": "推荐网站"},
		{"linkType": 2, "linkName": "个人网站"},
	}
	util.Success(c, links)
}

func GetLinkList(c *gin.Context) {
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "10"))

	links, total := model.GetAllLinks(page, limit)
	util.PageSuccess(c, links, total, page, limit)
}

func UpdateLinkStatus(c *gin.Context) {
	link := &model.Link{}
	if err := c.ShouldBindJSON(link); err != nil {
		util.Error(c, http.StatusBadRequest, "数据绑定失败")
		return
	}
	flag := model.UpdateLink(link)
	if flag {
		util.Success(c, link)
	} else {
		util.Error(c, http.StatusInternalServerError, "更新失败")
	}
}

func ClearLink(c *gin.Context) {
	id := c.Param("id")
	flag := model.DeleteLink(id)
	if flag {
		util.Success(c, id)
	} else {
		util.Error(c, http.StatusInternalServerError, "删除失败")
	}
}

func EditLink(c *gin.Context) {
	link := &model.Link{}
	if err := c.ShouldBind(link); err != nil {
		util.Error(c, http.StatusBadRequest, "数据绑定失败")
		return
	}
	link.CreatedAt = time.Now()
	if link.LinkId != "" {
		model.UpdateLink(link)
		util.Success(c, link.LinkId)
	} else {
		link.LinkId = util.GenerateUUID()
		flag := model.CreateLink(link)
		if flag {
			util.Success(c, link.LinkId)
		} else {
			util.Error(c, http.StatusInternalServerError, "添加失败")
		}
	}
}

func GetLinks(c *gin.Context) {
	favoriteLinks := model.GetLinksByType(0)
	recommendLinks := model.GetLinksByType(1)
	personalLinks := model.GetLinksByType(2)
	result := gin.H{
		"pageName":       "友情链接",
		"favoriteLinks":  favoriteLinks,
		"recommendLinks": recommendLinks,
		"personalLinks":  personalLinks,
	}
	util.Success(c, result)
}
